readthedocs/oauth/notifications.py

```diff
@@ -0,0 +1,38 @@
+# -*- coding: utf-8 -*-
+from __future__ import division, print_function, unicode_literals
+
+from readthedocs.notifications import Notification
+from readthedocs.notifications.constants import HTML, WARNING, INFO
+
+
+class AttachWebhookNotification(Notification):
```

Comment:
As I mentioned in the chat, I thought that we didn't use the `message_extends` directly, but through our Notification backend.

We can't create short-lived (non persistent) notifications from a Celery task since we don't have the request nor the session, which is the important part since the message is saved in the session.

> The package that I was using before, `django-async-messages`, saves the message in the cache instead of in the session and that's why we can save/load them without a request/session.

So, I'm still not sure what's the best approach here, but it looks like these are the main points:

* attaching the webhook fails: we want a persistent message here with an actionable link to solve the issue and mention the project's name because the user can moves out the project's page after importing it
* attaching the webhook succeeds: we want an async message saved in the cache

We could generalize this and create a celery notification helper that does one thing or the other depending on the status of the message (fail / success) so we don't have to deal with it every time we want a notification from a task.

Reply:
I think this needs some clarification, as it sounds like the notification system, and usage of message_extends, is confusing. We don't need to switch to a different package to do what you ultimately want to do.

message_extends is a) a databased backed method of sending messages, and is also b) a way to make persistent/sticky messages. It doesn't have to be both, it can accept message levels that are basic django message levels.

The message levels from `django.contrib.messages` correspond to the notification style you are used to -- messages that will display on one request only.

message_extends adds new levels that correspond to the same warning/info/etc levels, but provide a sticky/persistent message. This is a message that will display on multiple requests.

All of this is separate from how message_extends stores messages. You can pass any message level to message_extends and it will display, but it comes from the database, not the user's cookie/session.

Using our notification system isn't wrong, but it isn't required just to use message_extends. If you were to write this again, you don't need the bulk of our notification system. You would just submit a notification through message_extends -- and use a message level from `django.contrib.messages` to get a single request notification (non-sticky/non-persistent).

*However*, this is maybe a chance to improve a pattern for on-site messaging

I don't necessarily see a reason why we shouldn't use the notification system. My problem is that it is overkill for this case, but it does provide isolation between the notification pieces, which I like as pattern. Perhaps we need an extension to the class that makes it really easy for basic on-site messages like this. For example, a subclass that:

* Doesn't require files on disk for templates, it uses methods
* Doesn't send via the email backend, or maybe make this a class var that is default `False`
* Has success/failure messages in a single class -- normally this would be two separate classes, one for each notification and notification template
* Has a CBV like `get_success_url()` and `get_success_message()` or something, so we're not doing django templates in string vars on the class.

Just a thought. Hopefully that makes more sense. I'm happy to pair on this more if you're still unsure.

> We could generalize this and create a celery notification

We probably don't need this level of optimization yet, but the generalized on-site message notification helper class above would be close to this. We might be able to use that class to standardize all our on-site messaging eventually, celery task or not.